kvserver: Replace multiTestContext with TestCluster in client_replica_gc_test

lunevalex opened the pull request:
Makes progress on #8299

multiTestContext is legacy construct that is deprecated in favor of running
tests via TestCluster. This is one PR out of many to remove the usage of
multiTestContext in the client_replica_gc test cases.

Release note: none

Commit:
kvserver: Replace multiTestContext with TestCluster in client_replica_gc_test

Makes progress on #8299

multiTestContext is legacy construct that is deprecated in favor of running
tests via TestCluster. This is one PR out of many to remove the usage of
multiTestContext in the client_replica_gc test cases.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50657)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50657)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lunevalex and @tbg)

___
*[pkg/kv/kvserver/client_replica_gc_test.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-MB07X2i8VSUGc-L2ifH:-MB07X2i8VSUGc-L2ifI:bhmotiu) ([raw file](https://github.com/cockroachdb/cockroach/blob/44381670e7558f40bc83f71ce3d3582b3c57495c/pkg/kv/kvserver/client_replica_gc_test.go#L38)):*
> ```Go
> 	const numStores = 3
> 
> 	// Use actual engines (not in memory) because the in-mem ones don't write
> ```

I'm surprised we need this. I don't quite understand the justification. You say the test would still pass but then say it would probably fail. What were you seeing without it? The in-memory engine should still support a mock file system that every disk access should go through. We seem to handle this correctly below with calls like `fs.WriteFile(eng, ...). Are we forgetting to do that somewhere?
___
*[pkg/kv/kvserver/replica_gc_queue.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-MB09GcmC1jfQRJLtVzO:-MB09Gcn35fVrMcXke5a:b-wxs5oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/44381670e7558f40bc83f71ce3d3582b3c57495c/pkg/kv/kvserver/replica_gc_queue.go#L92)):*
> ```Go
> 	metrics                           ReplicaGCQueueMetrics
> 	db                                *kv.DB
> 	replicaGCQueueInactivityThreshold time.Duration
> ```

Are we actually using this? I would have expected to see something in `replicaGCShouldQueueImpl`.
___
*[pkg/kv/kvserver/replica_gc_queue.go, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-MB08bxz3QPxm7y5nojn:-MB08bxz3QPxm7y5nojo:b-vrgl61) ([raw file](https://github.com/cockroachdb/cockroach/blob/44381670e7558f40bc83f71ce3d3582b3c57495c/pkg/kv/kvserver/replica_gc_queue.go#L98)):*
> ```Go
> func newReplicaGCQueue(store *Store, db *kv.DB, gossip *gossip.Gossip) *replicaGCQueue {
> 	replicaGCQueueInactivityThreshold := ReplicaGCQueueInactivityThreshold
> 	if store.TestingKnobs().ReplicaGCQueueInactivityThreshold != 0 {
> ```

nit: this kind of thing is often done using a temp variable like:
```go
if knob := store.TestingKnobs().ReplicaGCQueueInactivityThreshold; knob != 0 {
	replicaGCQueueInactivityThreshold = knob
}
```


<!-- Sent from Reviewable.io -->

lunevalex commented:
___
*[pkg/kv/kvserver/replica_gc_queue.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-MB09GcmC1jfQRJLtVzO:-MB0LfJp5tFVno2wTfcZ:bez06od) ([raw file](https://github.com/cockroachdb/cockroach/blob/44381670e7558f40bc83f71ce3d3582b3c57495c/pkg/kv/kvserver/replica_gc_queue.go#L92)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are we actually using this? I would have expected to see something in `replicaGCShouldQueueImpl`.
</blockquote></details>

Good catch, I did not finish plumbing it through but the test passed. This was to replace the bit where were were manually moving the clock. It looks like its not needed because we disable the queue at the start of the cluster and then enable it, which means this gets picked up on the first run. Going to remove this. 


<!-- Sent from Reviewable.io -->

lunevalex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50657)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lunevalex, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/client_replica_gc_test.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-MB07X2i8VSUGc-L2ifH:-MB0EoYNEy5GhAZ2a514:b-48buck) ([raw file](https://github.com/cockroachdb/cockroach/blob/44381670e7558f40bc83f71ce3d3582b3c57495c/pkg/kv/kvserver/client_replica_gc_test.go#L38)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm surprised we need this. I don't quite understand the justification. You say the test would still pass but then say it would probably fail. What were you seeing without it? The in-memory engine should still support a mock file system that every disk access should go through. We seem to handle this correctly below with calls like `fs.WriteFile(eng, ...). Are we forgetting to do that somewhere?
</blockquote></details>

Great question, I did not change the logic or conditions of the test just it's usage of multiTestContext. I will test it out to see it makes a difference. 


<!-- Sent from Reviewable.io -->

lunevalex commented:
___
*[pkg/kv/kvserver/client_replica_gc_test.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-MB07X2i8VSUGc-L2ifH:-MB0OmbS3ZI1LTql1EYW:bj33hqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/44381670e7558f40bc83f71ce3d3582b3c57495c/pkg/kv/kvserver/client_replica_gc_test.go#L38)):*
<details><summary><i>Previously, lunevalex wrote…</i></summary><blockquote>

Great question, I did not change the logic or conditions of the test just it's usage of multiTestContext. I will test it out to see it makes a difference. 
</blockquote></details>

Actually you are right I am not sure why I thought the original test was creating file system based engines, I removed all of that and switched to in-memory ones and everything works. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks for the update, the PR is nice and straightforward now.

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50657#-:-MB0aefQ5sunt9xoWxOU:bkh8miv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

lunevalex commented:
bors r+
